nbgitpuller/plugins/plugin_helper.py

```diff
@@ -0,0 +1,146 @@
+        fname = fname[0].strip().strip('"')
+        ext = fname.split(".")[1]
+    elif file_type and "/zip" in file_type:
+        ext = "zip"
+    else:
+        url = url.split("/")[-1]
+        if "?" in url:
+            url = url[0:url.find('?')]
+        if "." in url:
+            ext = url.split(".")[1]
+
+    if not ext:
+        m = f"Could not determine the file extension for unarchiving: {url}"
+        raise Exception(m)
+    return ext
+
+
+# the downloaded content is in the response -- unarchive and save to the disk
+def save_response_content(url, response, temp_download_repo):
+    try:
+        ext = determine_file_extension(url, response)
+        CHUNK_SIZE = 32768
+        temp_download_file = f"{temp_download_repo}/download.{ext}"
+        with open(temp_download_file, "wb") as f:
+            for chunk in response.iter_content(CHUNK_SIZE):
+                # filter out keep-alive new chunks
+                if chunk:
+                    f.write(chunk)
+
+        shutil.unpack_archive(temp_download_file, temp_download_repo)
+
+        os.remove(temp_download_file)
+    except Exception as e:
+        m = f"Problem handling file download: {str(e)}"
+        raise Exception(m)
+
+
+# grab archive file from url
+def fetch_files(url, id=-1):
+    session = requests.Session()
+    session.mount('file://', FileAdapter())  # add adapter for pytests
+    response = session.get(url, params={'id': id}, stream=True)
+    token = get_confirm_token(response)
+    if token:
+        params = {'id': id, 'confirm': token}
+        response = session.get(url, params=params, stream=True)
+
+    return response
+
+
+# this drive the file handling -- called from zip_puller by all the
+# handle_files implementations for GoogleDrive, Dropbox, and standard
+# Web url
+def handle_files_helper(args):
+    try:
+        origin_repo = args["repo_parent_dir"] + args["origin_dir"]
+        temp_download_repo = args["repo_parent_dir"] + args["download_dir"]
+        if os.path.exists(temp_download_repo):
+            shutil.rmtree(temp_download_repo)
+
+        if not os.path.exists(origin_repo):
+            initialize_local_repo(origin_repo)
+
+        clone_local_origin_repo(origin_repo, temp_download_repo)
+        save_response_content(args["repo"], args["response"], temp_download_repo)
+        subprocess_helper("git add .", temp_download_repo)
+        subprocess_helper("git -c user.email=nbgitpuller@nbgitpuller.link -c user.name=nbgitpuller commit -m test --allow-empty", temp_download_repo)
+        subprocess_helper("git push origin master", temp_download_repo)
+        unzipped_dirs = os.listdir(temp_download_repo)
+
+        dir_names = list(filter(lambda dir: ".git" not in dir, unzipped_dirs))
+        return {"unzip_dir": dir_names[0], "origin_repo_path": origin_repo}
+    except Exception as e:
+        logging.exception(e)
+        raise ValueError(e)
+
+
+# executes git commands for us
+def subprocess_helper(cmd, cwd):
```

Comment:
We should use the array syntax throughout when calling external commands, especially when in a user facing program. Handling spaces in arguments - Like `echo "hi there"` or `echo hi\ there` or any number of variants is very difficult. It can also lead to security issues. Specifying them as a list of arrays fixes all that.

I'd suggest using `subprocess.check_output` - I think it does pretty much all the things you want here, and we can remove this helper

Reply:
yes got you here as well.